src/Processors/Sources/ShellCommandSource.cpp

```diff
@@ -353,7 +463,7 @@ namespace
 
                 if (!executor->pull(chunk))
                 {
-                    if (configuration.check_exit_code)
+                    if (check_exit_code)
                         command->wait();
                     return {};
                 }
```

Comment:
Here `ShellCommand::wait` works not as I would like - it mixes exit code with internal errors and throws an exception - it doesn't look correct to throw an exception for exit code inside this utility... but I think it's already impractical to refactor it now. So, probably we can catch the exception and rethrow it either it's not a `CHILD_WAS_NOT_EXITED_NORMALLY` or it is and reaction to exit code was configured. And by default I think reaction to exit code should be enabled.

Reply:
Yeah, we should handle exceptions in a fine-grained manner. By the way, we cannot check the exit code for pool variants.

Reply:
I think there is no need to distinguish different exceptions after second thoughts, because if `check_exit_code` is disabled, those internal errors will never happen anyway.